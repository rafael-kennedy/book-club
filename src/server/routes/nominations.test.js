const persistence = require("../persistence");
const supertest = require("supertest");
const app = require("../app");
const {
  validToken: token,
  adminToken
} = require("../../test/utils/fixture-tokens");
const { incomingPayload, seedData } = require("./nominations.fixture");

jest.setTimeout(60 * 1000);

let subjectApp;
let request;
let collection;

describe("nominations post route", () => {
  beforeAll(async () => {
    subjectApp = await app({ dbName: "test-nominations-post", listen: false });
    request = supertest(subjectApp);
    collection = subjectApp.get("db").collections.nominations;
  });
  test("can create new nomination", async () => {
    const result = await request
      .post("/nominations")
      .set("Authorization", "Bearer " + token)
      .send(incomingPayload())
      .expect(200);

    const inserted = await collection.findOne({ key: incomingPayload().key });
    expect(inserted).toBeTruthy();
    expect(inserted).toMatchSnapshot({
      _id: expect.anything()
    });
  });
});

describe("nominations get route", () => {
  beforeAll(async () => {
    subjectApp = await app({ dbName: "test-nominations-get", listen: false });
    request = supertest(subjectApp);
    collection = subjectApp.get("db").collections.nominations;
    await collection.insertMany(seedData());
  });
  test("can fetch my nominations", async () => {
    const result = await request
      .get("/nominations")
      .query({ creator: "me" })
      .send({ email: "test@test.com", password: "abcXYZ123!?" })
      .set("Authorization", "Bearer " + token)
      .expect(200);
    expect(result.body.length).toBe(2);
    expect(result.body).toMatchSnapshot();
  });

  test("can only fetch regular user's own nominations", async () => {
    const result = await request
      .get("/nominations")
      .send({ email: "test@test.com", password: "abcXYZ123!?" })
      .set("Authorization", "Bearer " + token)
      .expect(200);
    expect(result.body.length).toBe(2);
    expect(result.body).toMatchSnapshot();
  });

  test("admin can fetch all nominations", async () => {
    const result = await request
      .get("/nominations")
      .send({ email: "test@test.com", password: "abcXYZ123!?" })
      .set("Authorization", "Bearer " + adminToken)
      .expect(200);
    expect(result.body.length).toBe(3);
    expect(result.body).toMatchSnapshot();
  });
});
