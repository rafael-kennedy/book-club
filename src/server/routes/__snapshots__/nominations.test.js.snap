// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nominations get route admin can fetch all nominations 1`] = `
Array [
  Object {
    "_id": "5c06edc72ca07777df1dff64",
    "author": "Lee Child",
    "cover": "https://covers.openlibrary.org/w/id/240050-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/240050-L.jpg",
    "description": "Six shots.  and cunning—and then beat him shot for shot.",
    "key": "/works/OL52953W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2005,
    "title": "One Shot",
  },
  Object {
    "_id": "5c06f1ef004799790f896d33",
    "author": "Time Out",
    "cover": "https://covers.openlibrary.org/w/id/951934-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/951934-L.jpg",
    "description": "No description provided",
    "key": "/works/OL7971169W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2004,
    "title": "Time Out Madrid",
  },
  Object {
    "_id": "5c06f62d1739f87b9e8ffd1c",
    "author": "Robert A. Heinlein",
    "cover": "https://covers.openlibrary.org/w/id/6569996-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/6569996-L.jpg",
    "description": "Heinlein brought back Lazarus Long (his almost immortal) and covered parts of his life. It is one of Heinlein's best life, love, immortality and death.",
    "key": "/works/OL59694W",
    "nominatedBy": "5c05e8a02cfef52e5b197fbb",
    "published": 1973,
    "title": "Time Enough for Love",
  },
]
`;

exports[`nominations get route can fetch my nominations 1`] = `
Array [
  Object {
    "_id": "5c06edc72ca07777df1dff64",
    "author": "Lee Child",
    "cover": "https://covers.openlibrary.org/w/id/240050-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/240050-L.jpg",
    "description": "Six shots.  and cunning—and then beat him shot for shot.",
    "key": "/works/OL52953W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2005,
    "title": "One Shot",
  },
  Object {
    "_id": "5c06f1ef004799790f896d33",
    "author": "Time Out",
    "cover": "https://covers.openlibrary.org/w/id/951934-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/951934-L.jpg",
    "description": "No description provided",
    "key": "/works/OL7971169W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2004,
    "title": "Time Out Madrid",
  },
]
`;

exports[`nominations get route can only fetch regular user's own nominations 1`] = `
Array [
  Object {
    "_id": "5c06edc72ca07777df1dff64",
    "author": "Lee Child",
    "cover": "https://covers.openlibrary.org/w/id/240050-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/240050-L.jpg",
    "description": "Six shots.  and cunning—and then beat him shot for shot.",
    "key": "/works/OL52953W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2005,
    "title": "One Shot",
  },
  Object {
    "_id": "5c06f1ef004799790f896d33",
    "author": "Time Out",
    "cover": "https://covers.openlibrary.org/w/id/951934-S.jpg",
    "coverURL": "https://covers.openlibrary.org/w/id/951934-L.jpg",
    "description": "No description provided",
    "key": "/works/OL7971169W",
    "nominatedBy": "5c05e8a02cfef52e5b197fb5",
    "published": 2004,
    "title": "Time Out Madrid",
  },
]
`;

exports[`nominations post route can create new nomination 1`] = `
Object {
  "_id": Anything,
  "author": "Victoria Schwab",
  "cover": "https://covers.openlibrary.org/w/id/7410930-S.jpg",
  "coverURL": "https://covers.openlibrary.org/w/id/7410930-L.jpg",
  "description": "STEP INTO A UNIVERSE OF DARING ADVENTURE, THRILLING POWER, AND MULTIPLE LONDONS.",
  "key": "/works/OL17332803W",
  "nominatedBy": "5c05e8a02cfef52e5b197fb5",
  "published": 2015,
  "title": "A darker shade of magic",
}
`;
